File: Data Structures & Algorithms/permutations/submission-1.java
class Solution {

    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new LinkedList<>();

        if (nums.length == 0) {
            return res;
        }

        List<Integer> firstLis = new LinkedList<>();
        firstLis.add(nums[0]);
        res.add(firstLis);

        for (int i = 1; i < nums.length; i++) {
            List<List<Integer>> tempRes = new LinkedList<>();

            for (List<Integer> currList : res) {
                for (int j = 0; j < currList.size() + 1; j++) {
                    List<Integer> copyCurrList = new LinkedList<>(currList);
                    copyCurrList.add(j, nums[i]);
                    tempRes.add(copyCurrList);
                }
            }

            res = tempRes;
        }

        return res;
    }
}

File: Data Structures & Algorithms/combination-target-sum/submission-1.kt
class Solution {

    fun combinationSum(nums: IntArray, target: Int): List<List<Int>> {
        val finalResult: HashSet<List<Int>> = HashSet()
        combinationSum(nums, target, 0, LinkedList<Int>(), 0, finalResult)

        return finalResult.toList()
    }

    fun combinationSum(
        nums: IntArray,
        target: Int,
        currIdx: Int,
        currList: LinkedList<Int>,
        currSum: Int,
        finalResult: HashSet<List<Int>>
    ) {
        if (currIdx >= nums.size || currSum >= target) {
            if (currSum == target)
                finalResult.add(LinkedList(currList)) // n
            return
        }
        combinationSum(nums, target, currIdx + 1, currList, currSum, finalResult)
        currList.addLast(nums[currIdx])
        combinationSum(nums, target, currIdx, currList, currSum + nums[currIdx], finalResult)
        combinationSum(nums, target, currIdx + 1, currList, currSum + nums[currIdx], finalResult)
        currList.removeLast()
    }
}